src/lcm/LifeCycleStates.cc

```diff
@@ -2496,3 +2673,500 @@ void LifeCycleManager::trigger_resize_failure(int vid)
+        {
+            // FULL & backups > keep_last
+            ostringstream oss;
+
+            oss << "Removing backup snapshots:";
+
+            for (int i : delete_ids)
+            {
+                oss << " " << i;
+            }
+
+            vm->log("LCM", Log::INFO, oss.str());
+        }
+        else if (keep_last > 0 && increments > keep_last)
+        {
+            // INCREMENTAL & increments > keep_last
+            ostringstream oss;
+
+            oss << "Removing " << increments - keep_last << " backup increments";
+
+            vm->log("LCM", Log::INFO, oss.str());
+
+            //Flag the flatten operation to prevent race coditions with backup
+            backups.active_flatten(true);
+        }
+
+        vmpool->update(vm.get());
+
+        vm.reset();
+
+        /* ------------------------------------------------------------------ */
+        /* Add image to the datastore and forget keep_last backups            */
+        /* ------------------------------------------------------------------ */
+        if (mode == Backups::FULL || incremental_id == -1)
+        {
+            if ( auto ds = dspool->get(ds_id) )
+            {
+                ds->add_image(image_id);
+
+                dspool->update(ds.get());
+            }
+
+            for (int i : delete_ids)
+            {
+                if ( imagem->delete_image(i, error) != 0 )
+                {
+                    ostringstream oss;
+
+                    oss << "backup_success, cannot remove VM backup " << i
+                        << " : " << error;
+
+                    NebulaLog::error("LCM", oss.str());
+                }
+            }
+        }
+        else if (mode == Backups::INCREMENT && keep_last > 0 && increments > keep_last)
+        {
+            ostringstream oss;
+
+            oss << "<EXTRA_DATA>"
+                << "<KEEP_LAST>" << keep_last << "</KEEP_LAST>"
+                << "</EXTRA_DATA>";
+
+            if ( imagem->flatten_increments(image_id, ds_id, oss.str(), error) != 0 )
+            {
+                oss.str("");
+
+                oss << "backup_success, cannot flatten backup increments for image "
+                    << image_id << " : " << error;
+
+                NebulaLog::error("LCM", oss.str());
+            }
+        }
+
+        /* ------------------------------------------------------------------ */
+        /* Update quotas, count real size of the backup                       */
+        /* ------------------------------------------------------------------ */
+        ds_deltas.replace("SIZE", reserved_sz - real_sz);
+        ds_deltas.replace("IMAGES", 0);
+
+        Quotas::ds_del(vm_uid, vm_gid, &ds_deltas);
+
+        /* ------------------------------------------------------------------ */
+        /* Remove VM from Backup Job pending list                             */
+        /* ------------------------------------------------------------------ */
+        if (auto bj = bjpool->get(bj_id))
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This snippet may contain an accidental use of incorrect operation. Make sure that the operation being performed is in fact what you intended.